### PR TITLE
Issue3066

### PR DESCRIPTION
**Pull request by DrTodd13**
Solution for #3066.  Create right type for conversion of array.dtype to numpy.dtype(array_type).

### CODE REVIEW

File: numba/parfor.py [hunk 1]
```diff
@@ -587,18 +587,41 @@ def replace_func():
...
-                            self.typemap[typ_var.name] = types.DType(dtype)
+                            self.typemap[typ_var.name] = types.functions.NumberClass(dtype)
                             dtype_str = str(dtype)
                             if dtype_str == 'bool':
                                 dtype_str = 'bool_'
                             np_typ_getattr = ir.Expr.getattr(g_np_var, dtype_str, loc)
                             typ_var_assign = ir.Assign(np_typ_getattr, typ_var, loc)
-                            instr.value = typ_var
+
+                            dtype_attr_var = ir.Var(scope, mk_unique_var("$dtype_attr_var"), loc)
```
Comment (stuartarchibald) [inline numba/parfor.py]: I think a comment or two about the intent of this section of code would be useful, to include why it is being done and the NumPy equivalent expression it is representing?

File: numba/targets/npyimpl.py [hunk 2]
```diff
@@ -540,3 +541,47 @@ def generate(self, *args):
...
+        def codegen(context, builder, signature, args):
+            # All dtype objects are dummy values in LLVM.
+            # They only exist in the type level.
+            return context.get_dummy_value()
+
+        return sig, codegen
+
+@overload(np.dtype)
+def numpy_dtype(desc):
+    if isinstance(desc, types.Const):
```
Comment (stuartarchibald) [inline numba/targets/npyimpl.py]: Can these branches be combined like:
```
if isinstance(desc, (types.Const, types.functions.NumberClass)):
```
?

File: numba/targets/npyimpl.py [hunk 3]
```diff
@@ -540,3 +541,47 @@ def generate(self, *args):
...
+def _make_dtype_object(typingctx, desc):
+    """Given a string description *desc*, returns the dtype object.
+    """
+    if isinstance(desc, types.Const):
+        # Convert the str description into np.dtype then to numba type.
+        nb_type = from_dtype(np.dtype(desc.value))
+        return_type = types.DType(nb_type)
+        sig = return_type(desc)
+
+        def codegen(context, builder, signature, args):
```
Comment (stuartarchibald) [inline numba/targets/npyimpl.py]: Can these identical impls be lifted out of the condition?

File: numba/parfor.py (hunk 1 above)
Comment (DrTodd13) [reply numba/parfor.py]: Done.

File: numba/targets/npyimpl.py (hunk 2 above)
Comment (DrTodd13) [reply numba/targets/npyimpl.py]: Done.

File: numba/targets/npyimpl.py (hunk 3 above)
Comment (DrTodd13) [reply numba/targets/npyimpl.py]: Done.

**Comment by sklam** [comment]
Should add a test to check that the `str` path is tested; i.e.`np.dtype("int32")`.
The cov-diff is suggest that path is not exercised: https://codecov.io/gh/numba/numba/commit/603ea61387f143d736a77422703b7162dc128623#D1-562...563

**Comment by DrTodd13** [comment]
Haven't dealt with this code coverage reports before.  I assume this -4.46% is on a per-file basis?  Anyway, I added a couple dtype tests in test_types.py.  We'll see if that improves the coverage.

**Comment by stuartarchibald** [comment]
I think the giant difference in coverage is due to this: https://github.com/numba/numba/pull/3167. Looking at the pattern of files with coverage that has dropped suggests optional package support is the problem. When that's merged I'll restart the build test just to make sure.

**Comment by stuartarchibald** [comment]
@DrTodd13 I think the last thing to do ahead of merge is to add to the documentation that `np.dtype` is supported and that the `.type` attr on an instance works but the rest do not (here seems to be the place to add it): https://github.com/numba/numba/blob/ff1eaec52b659ee30e1fa0ff4920d06c60cb99c8/docs/source/reference/numpysupported.rst#L244

Thanks.

**Comment by stuartarchibald** [comment]
As I suspected, https://github.com/numba/numba/pull/3088#issuecomment-408020074, https://github.com/numba/numba/pull/3167 was the problem, checks pass now.

**Comment by DrTodd13** [comment]
Added that numpy.dtype is supported for its first argument.
Also added assertion that njit and parallel=True return types are identical.

**Comment by stuartarchibald** [comment]
Thanks for the fixes, 1e63917 is a valuable but weak test as e.g. `np.empty(1, np.float32)` and `np.empty(1, np.float64)` would be considered equal as the type is `np.ndarray`, however, I think enforcing strict type compliance between `njit(parallel=True/False)` is out of scope of this PR.

**Comment by stuartarchibald** [comment]
Thanks for writing this feature. CI has passed. Merging.

**Pull request merged**